base/sysinfo.jl

```diff
@@ -321,4 +323,81 @@ windows_version
 
 const WINDOWS_VISTA_VER = v"6.0"
 
+"""
+    Sys._can_execute(path::AbstractString)
+
+Returns `true` if the given `path` has executable permissions.
+"""
+function _can_execute(path::AbstractString)
+    # We use `access()` and `X_OK` to determine if a given path is executable
+    # by the current user.  `X_OK` comes from `unistd.h`.
+    X_OK = 1 << 0
+    @static if iswindows()
+        ccall(:_access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
+    else
+        ccall(:access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
+    end
+end
+
+"""
+    Sys.which(program_name::AbstractString)
+
+Given a program name, searches the current `PATH` to find the first binary with
+the proper executable permissions that can be run, and returns the absolute
+path. Raises `ArgumentError` if no such program is available.  If a relative or
+absolute path is passed in for `program_name`, that exact path is tested for
+executable permissions only, no searching of `PATH` is performed.
+"""
+function which(program_name::AbstractString)
+    # If prog has a slash, we know the user wants to determine whether the given
+    # file exists and is executable, and to not search the `PATH`.  Note that
+    # Windows can have either `\\` or `/` in its paths:
+    dirseps = @static if iswindows()
+        ["/", "\\"]
+    else
+        ["/"]
+    end
+
+    if any(occursin.(dirseps, program_name))
+        # If this file does not even exist, fail out
+        if !isfile(program_name)
+            throw(ArgumentError("$program_name does not exist"))
+        end
+
+        # If it does exist, check that it's executable or fail out
+        if !_can_execute(program_name)
+            throw(ArgumentError("$program_name is not executable"))
+        end
+
+        # If it all checks out, return the abspath
+        return abspath(program_name)
+    end
+
+    # If we have been given just a program name (not a relative or absolute
+    # path) then we should search `PATH` for it here:
+    path = get(ENV, "PATH", "")
+    pathsep = @static if iswindows()
+        ';'
+    else
+        ':'
+    end
+
+    # On windows, we need to check both $(program_name) and $(program_name).exe
+    program_matches(file, program_name) = @static if iswindows()
+        file == program_name || file == "$(program_name).exe"
```

Comment:
Don't you need to check everything in `get(ENV, "PATHEXT", "exe")`

Reply:
Also, you presumably only want to check appending an extension if `program_name` doesn't already have an extension.

Reply:
No to the first (also `.cmd`), yes to the second (with qualifications). We should probably try to return the same answer as our underlying spawn API (https://github.com/libuv/libuv/blob/edf05b97f03472def2837ceb2d6bf61a0d0dc248/src/win/process.c#L299)